File: android/BOINC/src/edu/berkeley/boinc/rpc/AcctMgrInfoParser.java
/*******************************************************************************
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2012 University of California
 * 
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 * 
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package edu.berkeley.boinc.rpc;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;

import edu.berkeley.boinc.utils.Logging;

import android.util.Log;
import android.util.Xml;

public class AcctMgrInfoParser extends BaseParser{

	private AcctMgrInfo mAcctMgrInfo = null;

	public AcctMgrInfo getAccountMgrInfo() {
		return mAcctMgrInfo;
	}

	public static AcctMgrInfo parse(String rpcResult) {
		try {
			AcctMgrInfoParser parser = new AcctMgrInfoParser();
			Xml.parse(rpcResult, parser);
			return parser.getAccountMgrInfo();
		} catch (SAXException e) {
			if(Logging.WARNING) Log.w(Logging.TAG,"AcctMgrRPCReplyParser: malformated XML");
			return null;
		}
	}

	@Override
	public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
		super.startElement(uri, localName, qName, attributes);
		if (localName.equalsIgnoreCase("acct_mgr_info")) {
			mAcctMgrInfo = new AcctMgrInfo();
		} else {
			mElementStarted = true;
			mCurrentElement.setLength(0);
		}
	}

	@Override
	public void endElement(String uri, String localName, String qName) throws SAXException {
		super.endElement(uri, localName, qName);
		try {
			if (mAcctMgrInfo != null) {
				// inside <acct_mgr_info>
				if (localName.equalsIgnoreCase("acct_mgr_info")) {
					// closing tag
					if (!mAcctMgrInfo.acct_mgr_name.isEmpty() && !mAcctMgrInfo.acct_mgr_url.isEmpty() && mAcctMgrInfo.have_credentials)
						mAcctMgrInfo.present = true;
				}
				else {
					// decode inner tags
					if (localName.equalsIgnoreCase("acct_mgr_name")) {
						mAcctMgrInfo.acct_mgr_name = mCurrentElement.toString();
					} else if (localName.equalsIgnoreCase("acct_mgr_url")) {
						mAcctMgrInfo.acct_mgr_url = mCurrentElement.toString();
					} else if (localName.equalsIgnoreCase("have_credentials")) {
						mAcctMgrInfo.have_credentials = true;
					} else if (localName.equalsIgnoreCase("cookie_required")) {
						mAcctMgrInfo.cookie_required = true;
					} else if (localName.equalsIgnoreCase("cookie_failure_url"))
						mAcctMgrInfo.cookie_failure_url = mCurrentElement.toString();
				}
			}
		} catch (Exception e) {}
		mElementStarted = false;
	}
}

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/AcctMgrInfoParser.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc

import android.util.Xml
import edu.berkeley.boinc.utils.Logging
import org.xml.sax.Attributes
import org.xml.sax.SAXException

class AcctMgrInfoParser : BaseParser() {
    lateinit var accountMgrInfo: AcctMgrInfo
        private set

    @Throws(SAXException::class)
    override fun startElement(uri: String?, localName: String, qName: String?, attributes: Attributes?) {
        super.startElement(uri, localName, qName, attributes)
        if (localName.equals(ACCT_MGR_INFO_TAG, ignoreCase = true)) {
            accountMgrInfo = AcctMgrInfo()
        } else {
            mElementStarted = true
            mCurrentElement.setLength(0)
        }
    }

    @Throws(SAXException::class)
    override fun endElement(uri: String?, localName: String, qName: String?) {
        super.endElement(uri, localName, qName)
        try {
            if (localName.equals(ACCT_MGR_INFO_TAG, ignoreCase = true)) { // closing tag
                if (arrayOf(accountMgrInfo.acctMgrName, accountMgrInfo.acctMgrUrl)
                                .none { it.isEmpty() } &&
                        accountMgrInfo.isHavingCredentials) {
                    accountMgrInfo.isPresent = true
                }
            } else { // decode inner tags
                when {
                    localName.equals(AcctMgrInfo.Fields.ACCT_MGR_NAME, ignoreCase = true) -> {
                        accountMgrInfo.acctMgrName = mCurrentElement.toString()
                    }
                    localName.equals(AcctMgrInfo.Fields.ACCT_MGR_URL, ignoreCase = true) -> {
                        accountMgrInfo.acctMgrUrl = mCurrentElement.toString()
                    }
                    localName.equals(AcctMgrInfo.Fields.HAVING_CREDENTIALS, ignoreCase = true) -> {
                        accountMgrInfo.isHavingCredentials = true
                    }
                }
            }
        } catch (e: Exception) {
            Logging.logException(Logging.Category.XML, "AcctMgrInfoParser.endElement error: ", e)
        }
        mElementStarted = false
    }

    companion object {
        const val ACCT_MGR_INFO_TAG = "acct_mgr_info"
        @JvmStatic
        fun parse(rpcResult: String?): AcctMgrInfo? {
            return try {
                val parser = AcctMgrInfoParser()
                Xml.parse(rpcResult, parser)
                parser.accountMgrInfo
            } catch (e: SAXException) {
                Logging.logException(Logging.Category.RPC, "AcctMgrInfoParser: malformed XML ", e)
                Logging.logDebug(Logging.Category.XML, "AcctMgrInfoParser: $rpcResult")

                null
            }
        }
    }
}
